### PR TITLE
Add ChannelInboundInvoker and ChannelOutboundInvoker

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

ChannelHandlerContext, ChannelPipeline and Channel share various method signatures. We should provide an interface to share.

Modifications:

Add ChannelInboundInvoker and ChannelOutboundInvoker and extend these.

Result:

Better API abstraction.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin PTAL ... This allows to more easily write impl that works with `Channel`, `ChannelHandlerContext` and `ChannelPipeline`.

**Comment by normanmaurer** [comment]
@Scottmitch as well ... PTAL

File: transport/src/main/java/io/netty/channel/AbstractChannel.java [hunk 1]
```diff
@@ -56,8 +56,6 @@
     private final ChannelId id;
     private final Unsafe unsafe;
     private final DefaultChannelPipeline pipeline;
-    private final ChannelFuture succeededFuture = new SucceededChannelFuture(this, null);
-    private final VoidChannelPromise voidPromise = new VoidChannelPromise(this, true);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: move this to DefaultChannelPipeline as we delegate for other methods into it as well

**Comment by Scottmitch** [comment]
lgtm

File: transport/src/main/java/io/netty/channel/AbstractChannel.java (hunk 1 above)
Comment (trustin) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: +1

**Comment by trustin** [comment]
LGTM. Nice change.

**Comment by normanmaurer** [comment]
@trustin thanks! Do we want the same for 4.0 but without adding the promise / future stuff to the ChannelOutboundInvoker interface?

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (27a392b87760a501f1519c0838b44194f236629f)

**Pull request closed without merging**